[FIX] crm: recompute probability for team-specific stages

pca-Odoo opened the pull request:
Before this commit, the Predictive Lead Scoring would calculate counts
for won / lost / total using the first global stage available. In
databases where there were no global stages, this resulted in zeros
for all variables which skipped writing an updated probability.

Behavior before fix:
1. Assign all crm stages to a team
2. Create a lead and move it to a won stage
-> Probability computed as 100% and set as won
3. Move lead back to a normal stage
-> Probability is not updated and remains as 100% and 'won'

This change will add an extra computation if the won / lost counts are
zeroes. While keeping using the current method (and potential overrides),
we will now fallback on the first stage of the team in that case. If
different, for instance when all stages are team-specific, or if first stage(s)
are team-specific, we will use those counts instead, if any. Behavior will not
be changed when the first stage overall is global, as in that case the first
stage of the team is that one.

opw-4573681


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/201256.png)](https://mergebot.odoo.com/odoo/odoo/pull/201256)

Odoonan commented on addons/crm/models/crm_lead.py:
```diff
@@ -2542,16 +2542,18 @@ def _pls_get_safe_fields(self):
         :return: won count, lost count and total count for all records in frequencies
         """
-        # TODO : check if we need to handle specific team_id stages [for lost count] (if first stage in sequence is team_specific)
         first_stage_id = self.env['crm.stage'].search([('team_id', '=', False)], order='sequence, id', limit=1)
+        if not first_stage_id and team_id > 0:
+            # For lost count, handle case when there are no global stages
+            first_stage_id = self.env['crm.stage'].search([('team_id','=', team_id)], order='sequence, id', limit=1)
```
Actually, we could save us a search here, and be even more explicit from a heuristics perspective.
I would probably go with something along those lines, to check of course :)

```python
stage_domain = [('team_id', '=', False)]
if team_id and team_id > 0:
    stage_domain = ['|', ('team_id', '=', False), ('team_id', '=', team_id)]
    # or stage_domain = [('team_id', 'in', [False, team_id])] ?
first_stage_id = self.env['crm.stage'].search(stage_domain, order='sequence, id', limit=1)
```

Odoonan replied:
This would probably need updating the last part of your commit message / PR message, as we would always pick the first among (global stages, team stages)

pca-Odoo replied:
Good catch, thank you

Odoonan commented on addons/crm/models/crm_lead.py:
```diff
@@ -2542,16 +2542,18 @@ def _pls_get_safe_fields(self):
                first stage can be used to know how many lost and won there is
                as won count are equals for all stage
                and first stage is always incremented in lost_count
         :param frequencies: lead_scoring_frequencies
```
Maybe add a small explanation for this new parameter (the rest of this docstring will be updated in master soon)

tde-banana-odoo replied:
Note that this is not stable compliant: if someone has an override, their module is broken.

Odoonan replied:
Ah yes, good catch. Maybe for stable, we could add team_id in the team_results before calling the method, use that value, and remove it once the call is done? (this method will probably disappear in master at some point, work in progress)

Odoonan replied:
Or, to avoid changing the content of the dict, we could instead do no modification in this function, but when the method returns 0,0,0 we do another check with team_id in the loop of _pls_get_naive_bayes_probabilities. This would probably be a totally stable-compliant solution :)

Odoonan reviewed (commented):
Hello @pca-Odoo ,

Thanks for the fix :)

Some work will be done soon in master, but in the meantime, 18.0 seems fine to me as proba will indeed never be (re)computed when all teams are assigned.

Left a few remarks , but the idea LGTM

Could you please add a small test in the pls part?

I would probably do something close to this:
- Ensure all stages belong to one team
- Create simple leads with that team in the first non-won stage
- Loose one of them, Win one of them (in order to fill the table)
- Rebuild frequency table
- Check new proba value of the second lead (low)
- Update proba of second lead by going to next non-won stage, check that probability indeed been changed (higher)

If I am not mistaken, the test should fail without your fix

Thanks and cheers!

pca-Odoo commented:
@Odoonan @tde-banana-odoo 
Thanks for the notes. I've reverted the method signature and implemented your change to avoid the second search.

I would like to get your thoughts on using a context instead? See updated code. My thought is to avoid copy/pasting the scoring calculation if the return is 0s.

I'll work out the tests soon, I'm a bit swamped for the week.
Thanks!

Odoonan commented on addons/crm/models/crm_lead.py:
```diff
@@ -2196,7 +2196,7 @@ def _pls_get_naive_bayes_probabilities(self, batch_mode=False):
             result[team_id]['team_won'], \
             result[team_id]['team_lost'], \
-            result[team_id]['team_total'] = self._pls_get_won_lost_total_count(result[team_id])
+            result[team_id]['team_total'] = self.with_context(pls_team_id=team_id)._pls_get_won_lost_total_count(result[team_id])
```
Hi @pca-Odoo ,

While more concise, we tend to avoid using context when we can. In this case the diff would not be much larger (even though there may be a bit of duplication) if we go with something like:
```python
team_won, team_lost, team_total = self._pls_get_won_lost_total_count(result[team_id])
if team_won == 0 and team_lost == 0 and team_total == 0 and team_id and team_id > 0:
    first_stage_id = self.env['crm.stage'].search([('team_id', '=', team_id)], order='sequence, id', limit=1)
    if str(first_stage_id.id) in team_results.get('stage_id', []):
      stage_result = team_results['stage_id'][str(first_stage_id.id)]
      team_won, team_lost, team_total = stage_result['won'], stage_result['lost'], stage_result['won'] + stage_result['lost']
```

Odoonan replied:
And then of course:
```python3
result[team_id]['team_won'], \
result[team_id]['team_lost'], \
result[team_id]['team_total'] = t_won, t_lost, t_total
```
Actually, let us rename to `t_won` `t_lost` `t_total` to avoid side effects with existing variables below

pca-Odoo commented on addons/crm/models/crm_lead.py:
```diff
@@ -2194,9 +2194,13 @@ def _pls_get_naive_bayes_probabilities(self, batch_mode=False):
+                if str(first_stage_id.id) in result[team_id].get('stage_id', []):
+                    stage_result = result[team_id]['stage_id'][str(first_stage_id.id)]
+                    t_won, t_lost, t_total = stage_result['won'], stage_result['lost'], stage_result['won'] + stage_result['lost']
+            result[team_id]['team_won'], result[team_id]['team_lost'], result[team_id]['team_total'] = t_won, t_lost, t_total
```
Can do. It felt silly deconstructing just to split the lines anyway.

pca-Odoo commented on addons/crm/models/crm_lead.py:
```diff
@@ -2194,9 +2194,13 @@ def _pls_get_naive_bayes_probabilities(self, batch_mode=False):
-            result[team_id]['team_total'] = self._pls_get_won_lost_total_count(result[team_id])
+            t_won, t_lost, t_total = self._pls_get_won_lost_total_count(result[team_id])
+            if t_won == 0 and t_lost == 0 and t_total == 0 and team_id and team_id > 0:
+                first_stage_id = self.env['crm.stage'].search([('team_id', '=', team_id)], order='sequence, id', limit=1)
```
Good catch

pca-Odoo commented:
@Odoonan done!

Odoonan commented on addons/crm/models/crm_lead.py:
```diff
@@ -2194,9 +2194,17 @@ def _pls_get_naive_bayes_probabilities(self, batch_mode=False):
         # Get all won, lost and total count for all records in frequencies per team_id
         for team_id in result:
+            t_won, t_lost, t_total = self._pls_get_won_lost_total_count(result[team_id])
+            # Fallback on team's first stage. Different if all stages are steam-specific. Clean me in master.
+            if t_won == 0 and t_lost == 0 and t_total == 0 and team_id and team_id > 0:
+                first_stage_id = self.env['crm.stage'].search([('team_id', 'in', [team_id, False])],
```
Hi @pca-Odoo ,

Thanks for the change, just spotted a tiny detail to improve left. Then we are good to go on my side :)

- typo `'steam-specific' -> 'team-specific'` (my bad :grin:) would not want to induce confusion here ^^
- While here, can you indent the first stage id as usual, for instance:
```python
first_stage_id = self.env['crm.stage'].search([
    ('team_id', 'in', [team_id, False])
], order='sequence, id', limit=1)
```


Cheers & thanks for the updates !

pca-Odoo replied:
Got it @Odoonan

pca-Odoo commented:
Hi @Odoonan 
Apologies for mia, had my hands full last week. Are there any changes still needed or can we deploy this fix? I just rebased and made sure tests are still passing

Odoonan commented:
Hello @tde-banana-odoo ,

Re-reading this small fix, I'd say it is ready, but I would not say no to a second opinion if you don't mind :) The solution here is to basically check the stage inside the method directly, instead of changing the signature of `_pls_get_won_lost_total_count`, which is not stable compliant as you mentioned.

Just to give some context, since it has been a while, the `team_id` values present in `result` come from the frequencies (or is '-1' to aggregate all results). As any won lead should update all the stages and create frequencies, we should not go into the additional code except in some corner-cases or in the case we are covering here: all stages are team-specific.

Advantages of the method:
- Any override of  `_pls_get_won_lost_total_count` will still be applied as the method is called first.
- We still use the frequency values linked to global stage if such stage exists, not changing the process in stable. (*)

Disadvantages of the method:
- A bit of duplication since we basically duplicate the content of `_pls_get_won_lost_total_count` 
- From a correctness perspective (*) , taking the first stage overall (between global and team-specific) makes more sense since the total of lost lead will be the one on that stage for the team. This is being changed in master, see WIP https://github.com/odoo/odoo/pull/159667/commits/21da729c6ddb740b31bf7aa2f3bb592838b889fb
       In order to do this instead in this fix, I see two ways:
      -> To not use `_pls_get_won_lost_total_count` at all and inline the stage computation (as in the wip).
      -> To pass the team_id to  `_pls_get_won_lost_total_count` with `with_context`, but AFAIK we do not like using context too much.

What do you think? Thanks for your input :pray: 

Cheers!
